File: codes/lua/clsobj/objmeth.lua
local Person = {}
Person.__index = Person

function Person:new(name, byear)    -- The constructor
    return setmetatable({name = name, byear = byear}, Person)
end

function Person:__tostring()
    return "(" .. self.name .. ": " .. self.byear .. ")"
end

function Person.__eq (p1, p2)
    return p1.name == p2.name
end

function Person.__lt (p1, p2)
    return (p1.byear < p2.byear) or (p1.byear == p2.byear and p1.name < p2.name)
end

function Person.__le (p1, p2)
    return (p1.byear < p2.byear) or (p1.byear == p2.byear and p1.name <= p2.name)
end

local p = Person:new("Karim", 1986)
local p2 = Person:new("Karim+1", 1986)
local p3 = {name = "Karim", byear = 1987}

print("As string: " .. tostring(p))
print("p == p3? " .. tostring(p==p3))
print("p ~= p3? " .. tostring(p~=p3))
print("p <= p2? " .. tostring(p <= p2))
print("p > p2? " .. tostring(p > p2))
print("p > p3? " .. tostring(p > p3))

File: codes/perl/clsobj/objmeth.pl
#! /usr/bin/perl -w

{
    package Person;

    sub new {
        my $class = shift;
        my $self = {
            _name => shift,
            _byear => shift,
        };
        bless $self, $class;
        return $self;
    }

    use overload
    '""' => sub {
        my ($self) = @_;
        return "($self->{_name}: $self->{_byear})"
    },
    '==' => sub {
        my ($self, $other) = @_;
        return ($self->{_name} eq $other->{_name})
    }

}

$p = new Person("Karim", 1986);
$p2 = new Person("Karim+1", 1986);
$p3 = {
    _name => "Karim",
    _byear => 1987,
};

sub boolval {
    my $bool = shift;
    return $bool? "true": "false";
}
print "As string: $p\n";
print "As string: " . $p2 . "\n";
print "p == p2: " . boolval($p == $p2) . "\n";
print "p == p3: " . boolval($p == $p3) . "\n";
